### PR TITLE
Skeletons & empty states

### PR DESCRIPTION
**Pull request by ilariaventurini**
Closes: #407.

Add empty states and skeletons.

We call **skeleton** the structure (with shimmer effect) created during the data loading and **empty state** the structure created when data are empty.

There are different types of skeletons (empty states) for the different graphs, as shown in the [issue #407](https://github.com/carbon-design-system/carbon-charts/issues/407).

There are 5 types of skeleton (empty state):
- with only horizontal lines
- with only vertical lines
- with both vertical and horizontal lines
- with a filled area (pie and donut).

In the demo page there are an empty state and a skeleton example for each type of chart. Some examples below:

horiz-empty

horiz-skeleton

vert-empty

vert-skeleton

grid-empty

grid-skeleton

pie-empty

pie-skeleton

donut-empty

donut-skeleton

Basically skeletons (and empty states) are composed by lines, rects and other shapes whose stroke/fill is an horizontal gradient. To achieve the shimmer effect, we animate (using d3) the `gradientTransform` translate attribute of the gradient.

### Review checklist (for reviewers only)
- [ ] Demos all features
- [ ] Documented/annotated
- [ ] Matches UI/UX specs
- [ ] Meets the [code style guide](https://github.com/IBM/carbon-charts/wiki/Code-style-guide)
- [ ] Accessible
- [ ] Mobile first (responsive)
- [ ] RTL support (bidirectional text)
- [ ] Performant (limited bloat)

### CODE REVIEW

**Comment by theiliad** [comment]
Thanks Ilaria 💯

Looking at the `core` storybook I don't see any skeleton states. Is that intended?

**Comment by ilariaventurini** [comment]
> Thanks Ilaria 💯
> 
> Looking at the `core` storybook I don't see any skeleton states. Is that intended?

You are right. 
For the moment, the skeletons are visible only in the charts inside the DEV stories.
I preferred to do so because I added a button that remove data (and then add them again) and I didn't want to "mess up" the other stories.

Let me know if you prefer to add the button in all the stories or if you have any other ideas :)

**Comment by theiliad** [comment]
> > Thanks Ilaria 💯
> > Looking at the `core` storybook I don't see any skeleton states. Is that intended?
> 
> You are right.
> For the moment, the skeletons are visible only in the charts inside the DEV stories.
> I preferred to do so because I added a button that remove data (and then add them again) and I didn't want to "mess up" the other stories.
> 
> Let me know if you prefer to add the button in all the stories or if you have any other ideas :)

I see, the demos look great!

Yes we should definitely add those buttons once the PR is ready

**Comment by ilariaventurini** [comment]
Hello everyone, I updated the PR description.

**Comment by caesarsol** [comment]
@theiliad we have several ways to include the test cases in the storybook, what option do you prefer?
 1. a button added to every chart in __DEV__ to see either the empty state or the loading state (as is now)
 2. a new page with all the skeletons for each chart
 3. in every chart section, two new sections for loading state and empty state

My favourite is 3, but @ilariaventurini argues that in that way we can't see an actual real-world representation of the `loading -> data arrives -> data disappears` transitions, and that's a good point as well.

**Comment by ilariaventurini** [comment]
> @theiliad we have several ways to include the test cases in the storybook, what option do you prefer?
> 
> 1. a button added to every chart in `__DEV__` to see either the empty state or the loading state (as is now)
> 2. a new page with all the skeletons for each chart
> 3. in every chart section, two new sections for loading state and empty state
> 
> My favourite is 3, but @ilariaventurini argues that in that way we can't see an actual real-world representation of the `loading -> data arrives -> data disappears` transitions, and that's a good point as well.

Hey @theiliad, what do you think about it? :)

**Comment by theiliad** [comment]
> > @theiliad we have several ways to include the test cases in the storybook, what option do you prefer?
> > 
> > 1. a button added to every chart in `__DEV__` to see either the empty state or the loading state (as is now)
> > 2. a new page with all the skeletons for each chart
> > 3. in every chart section, two new sections for loading state and empty state
> > 
> > My favourite is 3, but @ilariaventurini argues that in that way we can't see an actual real-world representation of the `loading -> data arrives -> data disappears` transitions, and that's a good point as well.
> 
> Hey @theiliad, what do you think about it? :)

I think @mjabbink can make the call here

**Comment by theiliad** [comment]
> > > @theiliad we have several ways to include the test cases in the storybook, what option do you prefer?
> > > 
> > > 1. a button added to every chart in `__DEV__` to see either the empty state or the loading state (as is now)
> > > 2. a new page with all the skeletons for each chart
> > > 3. in every chart section, two new sections for loading state and empty state
> > > 
> > > My favourite is 3, but @ilariaventurini argues that in that way we can't see an actual real-world representation of the `loading -> data arrives -> data disappears` transitions, and that's a good point as well.
> > 
> > 
> > Hey @theiliad, what do you think about it? :)
> 
> I think @mjabbink can make the call here

Let's do option 3 @ilariaventurini. Just add a button and down the road I'll switch that out for an entire toolbar to add more functionality 🙂

**Comment by ilariaventurini** [comment]
> > > > @theiliad we have several ways to include the test cases in the storybook, what option do you prefer?
> > > > 
> > > > 1. a button added to every chart in `__DEV__` to see either the empty state or the loading state (as is now)
> > > > 2. a new page with all the skeletons for each chart
> > > > 3. in every chart section, two new sections for loading state and empty state
> > > > 
> > > > My favourite is 3, but @ilariaventurini argues that in that way we can't see an actual real-world representation of the `loading -> data arrives -> data disappears` transitions, and that's a good point as well.
> > > 
> > > 
> > > Hey @theiliad, what do you think about it? :)
> > 
> > 
> > I think @mjabbink can make the call here
> 
> Let's do option 3 @ilariaventurini. Just add a button and down the road I'll switch that out for an entire toolbar to add more functionality 🙂

Hi @theiliad, sorry, I don't understand. Our idea of ​​option #3 does not include buttons. Can you explain better what you imagine the result is? :)

**Comment by ilariaventurini** [comment]
> > > > > @theiliad we have several ways to include the test cases in the storybook, what option do you prefer?
> > > > > 
> > > > > 1. a button added to every chart in `__DEV__` to see either the empty state or the loading state (as is now)
> > > > > 2. a new page with all the skeletons for each chart
> > > > > 3. in every chart section, two new sections for loading state and empty state
> > > > > 
> > > > > My favourite is 3, but @ilariaventurini argues that in that way we can't see an actual real-world representation of the `loading -> data arrives -> data disappears` transitions, and that's a good point as well.
> > > > 
> > > > 
> > > > Hey @theiliad, what do you think about it? :)
> > > 
> > > 
> > > I think @mjabbink can make the call here
> > 
> > 
> > Let's do option 3 @ilariaventurini. Just add a button and down the road I'll switch that out for an entire toolbar to add more functionality 🙂
> 
> Hi @theiliad, sorry, I don't understand. Our idea of ​​option #3 does not include buttons. Can you explain better what you imagine the result is? :)

@theiliad

**Comment by theiliad** [comment]
sorry @ilariaventurini, forgot to respond. your option 3 sounds good to me. would we be able to add those state under the `codesandbox` section? I'd still want codesandbox links to be easily found on the page.

as long as we make sure that the extra demos on each story are neatly organized, it works for me! 👍

**Comment by theiliad** [comment]
@designertyler please review https://user-images.githubusercontent.com/18484576/76979091-cce68a80-6937-11ea-9265-b4d9b164647f.png

**Comment by ilariaventurini** [comment]
Hi @theiliad, I updated the PR description. Let me know if everything is ok with you :)

File: packages/core/src/tools.ts [hunk 1]
```diff
@@ -250,7 +258,7 @@ export namespace Tools {
 	 * @returns The function to be used by D3 to push element to the top of the canvas
 	 */
 	export function moveToFront(element) {
-		return element.each(function() {
+		return element.each(function () {
```
Comment (theiliad) [inline packages/core/src/tools.ts:261]:
```suggestion
		return element.each(function() {
```

File: packages/core/src/components/graphs/bar-stacked.ts [hunk 2]
```diff
@@ -175,14 +175,14 @@ export class StackedBar extends Bar {
...
-			.on("click", function(datum) {
+			.on("click", function (datum) {
 				// Dispatch mouse event
 				self.services.events.dispatchEvent(Events.Bar.BAR_CLICK, {
 					element: select(this),
 					datum
 				});
 			})
-			.on("mouseout", function(datum) {
+			.on("mouseout", function (datum) {
```
Comment (theiliad) [inline packages/core/src/components/graphs/bar-stacked.ts:185]:
```suggestion
			.on("mouseout", function(datum) {
```

File: packages/core/src/components/graphs/bar-stacked.ts [hunk 3]
```diff
@@ -175,14 +175,14 @@ export class StackedBar extends Bar {
 					type: TooltipTypes.DATAPOINT
 				});
 			})
-			.on("click", function(datum) {
+			.on("click", function (datum) {
```
Comment (theiliad) [inline packages/core/src/components/graphs/bar-stacked.ts:178]:
```suggestion
			.on("click", function(datum) {
```

File: packages/core/src/components/graphs/bar-stacked.ts [hunk 4]
```diff
@@ -134,7 +134,7 @@ export class StackedBar extends Bar {
 
 		const self = this;
 		this.parent.selectAll("path.bar")
-			.on("mouseover", function(datum) {
+			.on("mouseover", function (datum) {
```
Comment (theiliad) [inline packages/core/src/components/graphs/bar-stacked.ts:137]:
```suggestion
			.on("mouseover", function(datum) {
```

File: packages/core/src/components/axes/zero-line.ts [hunk 5]
```diff
@@ -7,12 +7,29 @@ export class ZeroLine extends Component {
 	type = "zero-line";
 
 	render(animate: boolean) {
+		const [minDomainValue, maxDomainValue] = this.services.cartesianScales.getRangeScale().domain();
+		const containsZero = (minDomainValue * maxDomainValue) < 0;
```
Comment (theiliad) [inline packages/core/src/components/axes/zero-line.ts:11]: hmmm... would this be true if domain is [0, 200]? do we care about that case?

File: packages/core/src/components/axes/grid.ts [hunk 6]
```diff
@@ -80,6 +94,10 @@ export class Grid extends Component {
 
 		// find the 2 gridlines on either side of the mouse
 		let floor = -1; let ceiling;
```
Comment (theiliad) [inline packages/core/src/components/axes/grid.ts:96]:
```suggestion
		let floor = -1;
		let ceiling;
```

File: packages/core/src/components/axes/axis.ts [hunk 7]
```diff
@@ -293,28 +306,28 @@ export class Axis extends Component {
...
-			.on("click", function(datum) {
+			.on("click", function (datum) {
 				// Dispatch mouse event
 				self.services.events.dispatchEvent(Events.Axis.LABEL_CLICK, {
 					element: select(this),
 					datum
 				});
 			})
-			.on("mouseout", function(datum) {
+			.on("mouseout", function (datum) {
```
Comment (theiliad) [inline packages/core/src/components/axes/axis.ts:330]:
```suggestion
			.on("mouseout", function(datum) {
```

File: packages/core/src/components/axes/axis.ts [hunk 8]
```diff
@@ -293,28 +306,28 @@ export class Axis extends Component {
...
-			.on("mousemove", function(datum) {
+			.on("mousemove", function (datum) {
 				// Dispatch mouse event
 				self.services.events.dispatchEvent(Events.Axis.LABEL_MOUSEMOVE, {
 					element: select(this),
 					datum
 				});
 			})
-			.on("click", function(datum) {
+			.on("click", function (datum) {
```
Comment (theiliad) [inline packages/core/src/components/axes/axis.ts:323]:
```suggestion
			.on("click", function(datum) {
```

File: packages/core/src/components/axes/axis.ts [hunk 9]
```diff
@@ -293,28 +306,28 @@ export class Axis extends Component {
...
-			.on("mouseover", function(datum) {
+			.on("mouseover", function (datum) {
 				// Dispatch mouse event
 				self.services.events.dispatchEvent(Events.Axis.LABEL_MOUSEOVER, {
 					element: select(this),
 					datum
 				});
 			})
-			.on("mousemove", function(datum) {
+			.on("mousemove", function (datum) {
```
Comment (theiliad) [inline packages/core/src/components/axes/axis.ts:316]:
```suggestion
			.on("mousemove", function(datum) {
```

File: packages/core/src/components/axes/axis.ts [hunk 10]
```diff
@@ -293,28 +306,28 @@ export class Axis extends Component {
 
 		const self = this;
 		container.selectAll("g.tick text")
-			.on("mouseover", function(datum) {
+			.on("mouseover", function (datum) {
```
Comment (theiliad) [inline packages/core/src/components/axes/axis.ts:309]:
```suggestion
			.on("mouseover", function(datum) {
```

File: packages/core/demo/data/bar.ts [hunk 11]
```diff
@@ -215,9 +215,46 @@ export const stackedBarTimeSeriesData = [
...
+// 	{ group: "Dataset 3", date: new Date(2019, 0, 5), value: 15000 },
+// 	{ group: "Dataset 3", date: new Date(2019, 0, 8), value: 20000 },
+// 	{ group: "Dataset 3", date: new Date(2019, 0, 13), value: 39213 },
+// 	{ group: "Dataset 3", date: new Date(2019, 0, 17), value: 61213 },
+// 	{ group: "Dataset 4", date: new Date(2019, 0, 2), value: 10 },
+// 	{ group: "Dataset 4", date: new Date(2019, 0, 6), value: 37312 },
+// 	{ group: "Dataset 4", date: new Date(2019, 0, 8), value: 51432 },
+// 	{ group: "Dataset 4", date: new Date(2019, 0, 15), value: 40323 },
+// 	{ group: "Dataset 4", date: new Date(2019, 0, 19), value: 31300 },
+// ];
```
Comment (theiliad) [inline packages/core/demo/data/bar.ts:256]: do we need this?

File: packages/core/src/components/graphs/skeleton.ts [hunk 12]
```diff
@@ -0,0 +1,246 @@
...
+import { Skeletons } from "../../interfaces/enums";
+
+export class Skeleton extends Component {
+	type = "skeleton";
+	xScale: any;
+	yScale: any;
+	backdrop: any;
+
+	render() {
+		const areDataEmpty = this.model.areDataEmpty();
```
Comment (theiliad) [inline packages/core/src/components/graphs/skeleton.ts:18]:
```suggestion
		const isDataEmpty = this.model.isDataEmpty();
```

File: packages/core/src/components/graphs/skeleton.ts [hunk 13]
```diff
@@ -0,0 +1,246 @@
...
+		const isDataLoading = this.model.getOptions().data.loading;
+
+		if (areDataEmpty) {
+			this.renderSkeleton(isDataLoading);
+		} else if (!areDataEmpty && isDataLoading) {
+			throw new Error(`There's something wrong. You can't have not empty data and data loading togheter.`);
```
Comment (theiliad) [inline packages/core/src/components/graphs/skeleton.ts:24]:
```suggestion
			throw new Error(`Something went wrong. You can't provided non-empty data and data loading together.`);
```

File: packages/core/src/components/graphs/skeleton.ts [hunk 14]
```diff
@@ -0,0 +1,246 @@
...
+		}
+	}
+
+	renderSkeleton(showShimmerEffect: boolean) {
+		if (this.configs.skeleton === Skeletons.GRID) {
+			this.renderSkeletonGrid(showShimmerEffect);
+		} else if (this.configs.skeleton === Skeletons.VERT_OR_HORIZ) {
+			this.renderSkeletonVertOrHoriz(showShimmerEffect);
+		} else if (this.configs.skeleton === Skeletons.PIE) {
+			this.renderSkeletonPie(showShimmerEffect);
```
Comment (theiliad) [inline packages/core/src/components/graphs/skeleton.ts:36]:
```suggestion
			this.renderPieSkeleton(showShimmerEffect);
```

File: packages/core/src/components/graphs/skeleton.ts [hunk 15]
```diff
@@ -0,0 +1,246 @@
...
+		} else {
+			this.removeSkeleton();
+		}
+	}
+
+	renderSkeleton(showShimmerEffect: boolean) {
+		if (this.configs.skeleton === Skeletons.GRID) {
+			this.renderSkeletonGrid(showShimmerEffect);
+		} else if (this.configs.skeleton === Skeletons.VERT_OR_HORIZ) {
+			this.renderSkeletonVertOrHoriz(showShimmerEffect);
```
Comment (theiliad) [inline packages/core/src/components/graphs/skeleton.ts:34]:
```suggestion
			this.renderVertOrHorizSkeleton(showShimmerEffect);
```

File: packages/core/src/components/graphs/skeleton.ts [hunk 16]
```diff
@@ -0,0 +1,246 @@
...
+		} else if (!areDataEmpty && isDataLoading) {
+			throw new Error(`There's something wrong. You can't have not empty data and data loading togheter.`);
+		} else {
+			this.removeSkeleton();
+		}
+	}
+
+	renderSkeleton(showShimmerEffect: boolean) {
+		if (this.configs.skeleton === Skeletons.GRID) {
+			this.renderSkeletonGrid(showShimmerEffect);
```
Comment (theiliad) [inline packages/core/src/components/graphs/skeleton.ts:32]:
```suggestion
			this.renderGridSkeleton(showShimmerEffect);
```

File: packages/core/src/components/graphs/skeleton.ts [hunk 17]
```diff
@@ -0,0 +1,246 @@
...
+		} else if (this.configs.skeleton === Skeletons.VERT_OR_HORIZ) {
+			this.renderSkeletonVertOrHoriz(showShimmerEffect);
+		} else if (this.configs.skeleton === Skeletons.PIE) {
+			this.renderSkeletonPie(showShimmerEffect);
+		} else if (this.configs.skeleton === Skeletons.DONUT) {
+			this.renderSkeletonDonut(showShimmerEffect);
```
Comment (theiliad) [inline packages/core/src/components/graphs/skeleton.ts:38]:
```suggestion
			this.renderDonutSkeleton(showShimmerEffect);
```

File: packages/core/src/components/graphs/skeleton.ts [hunk 18]
```diff
@@ -0,0 +1,246 @@
...
+		if (showShimmerEffect) {
+			this.setShimmerEffect("shimmer-lines");
+		}
+	}
+
+	renderSkeletonVertOrHoriz(showShimmerEffect: boolean) {
+		const orientation = this.services.cartesianScales.getOrientation();
+		this.setScales();
+		this.drawBackdrop(showShimmerEffect);
+		if (orientation === "vertical") {
```
Comment (theiliad) [inline packages/core/src/components/graphs/skeleton.ts:58]:
```suggestion
		if (orientation === CartesianOrientations.VERTICAL) {
```

File: packages/core/src/components/graphs/skeleton.ts [hunk 19]
```diff
@@ -0,0 +1,246 @@
...
+			this.drawYGrid(showShimmerEffect);
+		}
+		if (orientation === "horizontal") {
```
Comment (theiliad) [inline packages/core/src/components/graphs/skeleton.ts:61]:
```suggestion
		if (orientation === CartesianOrientations.HORIZONTAL) {
```

File: packages/core/src/components/axes/zero-line.ts (hunk 5 above)
Comment (ilariaventurini) [reply packages/core/src/components/axes/zero-line.ts:11]: If domain is `[0, 200]`, `containsZero` will be false and I think that in that case is ok to not show the zero axis line because it coincides with the x axis.
But I can make it clearer replacing `(minDomainValue * maxDomainValue) < 0` with `minDomainValue > 0 && maxDomainValue < 0 || minDomainValue < 0 && maxDomainValue > 0`

Comment (theiliad) [reply packages/core/src/components/axes/zero-line.ts:11]: ok, technically the domain contains a zero at that point, so should we rename `containsZero` to `drawZeroLine`?

Comment (ilariaventurini) [reply packages/core/src/components/axes/zero-line.ts:11]: I absolutely agree with you

File: packages/core/src/styles/components/_skeleton.scss [hunk 20]
```diff
@@ -0,0 +1,57 @@
+$lines-bg-color: $ui-03;
+$lines-shimmer-color: white;
+$areas-bg-color: rgba(127, 127, 127, 0.1);
+$areas-shimmer-color: rgba(255, 255, 255, 0.15);
```
Comment (natashadecoste) [inline packages/core/src/styles/components/_skeleton.scss:4]:
```suggestion
$areas-shimmer-color: rgba($carbon--white-0, 0.15);
```

File: packages/core/src/components/graphs/skeleton.ts [hunk 21]
```diff
@@ -0,0 +1,246 @@
...
+
+		backdropRect
+			.classed("shimmer-effect-lines", showShimmerEffect)
+			.classed("empty-state-lines", !showShimmerEffect);
+	}
+
+	drawXGrid(showShimmerEffect: boolean) {
+		const height = this.backdrop.attr("height");
+		const width = this.backdrop.attr("width");
+		const ticksNumber = 5;
```
Comment (natashadecoste) [inline packages/core/src/components/graphs/skeleton.ts:119]: do we want this in configuration so it can be changed by users? and then set 5 as the default

File: packages/core/src/components/graphs/skeleton.ts [hunk 22]
```diff
@@ -0,0 +1,246 @@
...
+
+		xGridG.selectAll("line")
+			.classed("shimmer-effect-lines", showShimmerEffect)
+			.classed("empty-state-lines", !showShimmerEffect);
+	}
+
+	drawYGrid(showShimmerEffect: boolean) {
+		const height = this.backdrop.attr("height");
+		const width = this.backdrop.attr("width");
+		const ticksNumber = 5;
```
Comment (natashadecoste) [inline packages/core/src/components/graphs/skeleton.ts:140]: same as above

File: packages/core/src/components/graphs/skeleton.ts (hunk 21 above)
Comment (ilariaventurini) [reply packages/core/src/components/graphs/skeleton.ts:119]: image

I think we could reuse `numberOfTicks` from` grid`, what do you think?

**Comment by ilariaventurini** [comment]
> In firefox the skeleton charts dont seem to render at all in the storybook

I'll take a look. Thank you

File: packages/core/src/components/graphs/skeleton.ts (hunk 21 above)
Comment (caesarsol) [reply packages/core/src/components/graphs/skeleton.ts:119]: I think it's a bit much to let users customize how many ticks to show for a skeleton, IMO. :)
What about a constant in the file?

**Comment by ilariaventurini** [comment]
> > In firefox the skeleton charts dont seem to render at all in the storybook
> 
> I'll take a look. Thank you

@natashadecoste fixed.
